invalidate_request.py

```diff
@@ -0,0 +1,417 @@
+"""
+This module allows to invalidate a 
+root requests and its generated chains.
+This is intended to be used only for
+`production_manager` users or `administrators`.
+"""
+
+# Add path to code that allow easy access to McM
+# sys.path.append("/afs/cern.ch/cms/PPD/PdmV/tools/McM/")
+
+import datetime
+import logging
+import pprint
+import sys
+from itertools import groupby
+from copy import deepcopy
+from rest import McM
+
+# Check if version is >= 3.11, otherwise raise an exception.
+if sys.version_info < (3, 11):
+    msg = (
+        "Unsuported Python version.\n"
+        "Please use a version equal of higher than Python 3.11.\n"
+        "Current version: %s " % (sys.version)
+    )
+    raise RuntimeError(msg)
+
+
+class InvalidateDeleteRequests:
+    """
+    Invalidate and delete all the chain requests in McM
+    linked to a root request in McM and delete the
+    root requests if required.
+    """
+
+    def __init__(self, mcm: McM) -> None:
+        self.mcm = mcm
+        self.logger = self._get_logger()
+        self._chain_request_type = "chained_requests"
+        self._request_type = "requests"
+        self.logger.warning("Sending requests to target environment: %s", self.mcm.host)
+
+    def _get_logger(self) -> logging.Logger:
+        """
+        Creates a logger to record the performed steps.
+        """
+        logger: logging.Logger = logging.getLogger(__name__)
+        formatter = logging.Formatter("[%(asctime)s][%(levelname)s] %(message)s")
+        current_date = str(datetime.datetime.now().strftime("%Y_%m_%d_%H_%M"))
+        fh: logging.Handler = logging.FileHandler(
+            f"mcm_invalidate_requests_{current_date}.log"
+        )
+        fh.setFormatter(formatter)
+        logger.addHandler(fh)
+        return logger
+
+    def _pretty(self, obj: dict) -> str:
+        """
+        Pretty print an object in console
+        """
+        return pprint.pformat(obj, width=50, compact=True)
+
+    def _get(self, endpoint: str) -> dict:
+        """
+        Execute a raw GET operation for a resource in McM.
+        """
+        return self.mcm._McM__get(endpoint)
```

Comment:
If the function should be public, make it public in the original class. Do not bypass name mangling.

Reply:
Agree with this but let’s defer it to a future update. This would require performing changes in the `rest.py` module, this module is loaded dynamically in scripts of our user's code and I have seen some examples when the mangled method is called to perform another kind of operation. If we perform that change now, there is going to be broken code we’re now even aware of.

Reply:
I'd argue that using private API is unsupported and changes can be done without notice. Or you can just add a public method that exposes the private one.

Reply:
Ideally, we should be able to perform it but consider this case [CMS GEN – request_fragment_check.py](https://github.com/cms-sw/genproductions/blob/master/bin/utils/request_fragment_check.py#L100), a mangled method is used because of the absence of a public method to perform a “raw” GET request. Also, related to the McM REST client module, the HTTP methods functions (like `get(...)`, `put(...)`) are tied to specific kinds of queries and the `private` API. Because of the last, we can also find other mangled methods being used to perform queries to the public API for the example described. In conclusion, changing the current implementation now is not feasible, and creating another method to wrap it like `raw_get(...)` seems ugly to me. Also, this change would be out of the scope of this PR.

Nevertheless, I share your concern about the code quality for this tool and that it should be refactored. I have other ideas related to this, they are documented in the following issue: [#23](https://github.com/cms-PdmV/mcm_scripts/issues/23).

Reply:
OK.